### DISCUSSION TITLE
Allowing for custom REFPROP installation directory using the Python wrapper

### DISCUSSION THREAD

**Discussion by zabidi17** [usnistgov/REFPROP-issues#586] [Q&A] [answered]
Hello,

I'm using REFPROP to provide fluid properties for a calculation tool. Due to discrepancies with REFPROP installation location across different machines in my department, I need to allow a user to provide a custom directory for REFPROP.  This question may be more of a Python wrapper question, and my difficulties may be because I'm not implementing the wrapper appropriately. I initialize REFPROP in a separate file from where I actually call the wrapper using the following code:
```
from ctREFPROP.ctREFPROP import REFPROPFunctionLibrary
import os
from pathlib import Path

# def ref_start(file = ""):
#     if not file:
#         os.environ['RPPREFIX'] = r'C:/Program Files (x86)/REFPROP'
#     else:
#         os.environ['RPPREFIX'] = file

#     RP = REFPROPFunctionLibrary(os.environ['RPPREFIX'])

#     RP.SETPATHdll(os.environ['RPPREFIX'])

#     MASS_BASE_SI = RP.GETENUMdll(0, "MASS BASE SI").iEnum


def density(fluid, firstValue, secondValue, MASS_BASE_SI, inputType):
    result = RP.REFPROPdll(fluid, inputType, "D", MASS_BASE_SI, 0, 0 , firstValue, secondValue, [1.0])
    return result.Output[0]

def enthalpy(fluid, firstValue, secondValue, MASS_BASE_SI, inputType):
    result = RP.REFPROPdll(fluid, inputType, "H", MASS_BASE_SI, 0, 0 , firstValue, secondValue, [1.0])
    return result.Output[0]
    

os.environ['RPPREFIX'] = r'C:/Program Files (x86)/REFPROP'
RP = REFPROPFunctionLibrary(os.environ['RPPREFIX'])

RP.SETPATHdll(os.environ['RPPREFIX'])

MASS_BASE_SI = RP.GETENUMdll(0, "MASS BASE SI").iEnum
```

I had some issues with this code structure before, but this seems to work. I've tried to have the REFPROP initialization inside a function where I can provide an optional filename, but this causes issues with all of my other functions due to the "RP" object being called before initialization. From my understanding, REFPROP will be initialized as soon as this module is imported but I would need to apply some sort of condition and pass some information to this module to allow for a custom directory.

I'm not sure how to go about solving this. Is initializing REFPROP from a separate file not a good idea due for this case?

**Comment by ianhbell** [comment]
I don't understand what the problem is. You can point REFPROPFunctionLibrary at a particular shared library, and then call SETPATHdll function with a different location (which could be a shared folder, but this is NOT(!) recommended due to the risk of blocking file reads from multiple users.

**Comment by ianhbell** [comment]
Also, your code you actually run should be inside a 
``` python
if __name__ == '__main__':
    ....
``` 
block, and initialization and calls should happen inside this block. It should work fine, I use this approach all the time with this interface. If not, please provide a runnable(!) example.

**Reply by zabidi17** [reply]
So I'm not sure if this is what I would need because I am intending to import functions from a module rather than executing this module as a standalone file.

I can expand on what I'm trying to do, because I realize I'm probably not explaining it clearly.

I have a file for running a GUI, for now lets call that "gui.py". My goal within "gui.py" is to ask the user for a custom REFPROP directory then some how pass that a function that will initialize the REFPROP instance. Then, "gui.py" performs some calculations that involve REFPROP queries, but these functions are defined in a separate module, "calculations.py".

Here's some code that illustrates what I would need from "gui.py"
```
# gui.py
from tkinter import *
from tkinter.simpledialog import askstring
from ref_start_test import *
from calculate import *

root = Tk()
root.geometry('800x880')

# Prompt the user for the directory
refprop_custom_dir = askstring('REFPROP Location', 'Enter Custom REFPROP File Location:')

# pass the custom directory to the initialization function
RP, MASS_BASE_SI = start_refprop(refprop_custom_dir)

mdot = calc(RP, MASS_BASE_SI, 5, 2)

```

So in "gui.py", the custom directory information is received and the REFPROP initialization is done. This file imports the "start_refprop" function from "refprop_start.py"
which is the following:
```
# refprop_start.py
from ctREFPROP.ctREFPROP import REFPROPFunctionLibrary
import os
from pathlib import Path

def start_refprop(file = 0):
    if not file:
        os.environ['RPPREFIX'] = r'C:/Program Files (x86)/REFPROP'
        RP = REFPROPFunctionLibrary(os.environ['RPPREFIX'])
    else:
        os.environ['RPPREFIX'] = file

    RP = REFPROPFunctionLibrary(os.environ['RPPREFIX'])

    RP.SETPATHdll(os.environ['RPPREFIX'])
    print(RP.RPVersion())
    MASS_BASE_SI = RP.GETENUMdll(0, "MASS BASE SI").iEnum
    return(RP, MASS_BASE_SI)


def density(fluid, firstValue, secondValue, MASS_BASE_SI, inputType):
    result = RP.REFPROPdll(fluid, inputType, "D", MASS_BASE_SI, 0, 0 , firstValue, secondValue, [1.0])
    return result.Output[0]

def enthalpy(fluid, firstValue, secondValue, MASS_BASE_SI, inputType):
    result = RP.REFPROPdll(fluid, inputType, "H", MASS_BASE_SI, 0, 0 , firstValue, secondValue, [1.0])
    return result.Output[0]

```

And a third module includes various functions for calculations. here's an example code with a small function using the custom REFPROP density wrapper function to calculate flow rate. 

```
# calculations.py
from ref_start_test import *


def calc(RP, MASS_BASE_SI, v, a):
    
    rho = density("Water", "25","101325", MASS_BASE_SI, "TP")

    return rho*v*a

```

I don't think I'm setting up my file structure, scoping, and function definition properly. I keep having various issues due to the "RP" object not being defined during the definition of the density() function in refprop_start. I've also run into issues where the "calc" function fails when trying to execute the REFPROP call. I'm realizing while typing this out that I may want to separate the definitions of the wrapper functions ( density(), enthalpy() ) and the refprop_start() function into separate modules to avoid a scoping error, but I thought I'd leave my current implementation and get your advice on what would be the best way to structure my code for this situation.

I appreciate your quick response and I apologize if my initial question was vague. I hope this better explains what I'm looking for.

**Comment by ianhbell** [comment]
A core comment is that REFPROP is not thread-safe, so you could get surprising behavior in a GUI that uses multiple threads

**Comment by ianhbell** [comment, marked as answer]
I would instantiate the copy in your main thread, and pass that to other threads

**Reply by zabidi17** [reply]
After reading this and your other comment on thread-safety, it clicked for me, and I realized how I should be structuring my code. Now I'm initializing REFPROP alongside my GUI and passing along that instance to wherever I plan to use it, just as you suggested. I also was having issues with my customer wrapper functions because I wasn't passing the REFPROP instance as an argument. Thanks for the help, and sorry about being vague!

**Comment by ianhbell** [comment]
See for instance how I handled that in the code from this paper: https://www.sciencedirect.com/science/article/pii/S0140700719302361

**Reply by zabidi17** [reply]
Quick question. In this example you create an instance of REFPROP in every function where you use it. I assumed this would cause issues with having multiple instances but I'm realizing that python garbage collector would handle discarding the unused instances, is that correct?

**Reply by ianhbell** [reply]
Depending on exactly how you handle things it should be ok. In that case, the calls are happening in separate PROCESSES, so they are safe, each process gets its own instance of REFPROP, and so long as the file reads don't clash, you should be good to go.
